### PR TITLE
[mlir][EmitC] Refactor MLIR Translate To Cpp

### PR DESCRIPTION
**Pull request by weizchen**
Hi, this aims to refactor the MLIR TranslateToCpp part by introducing CppTranslationInterface that breaks down the overall translation into separate dialect cpp translations. It follows the way that LLVMTranslation is constructed in MLIR. 

I believe this improves the overall readability of code structure while could help with the extensibility for future c++ translation addition like using vector dialect to translate to ACLE or other vector extensions for X86 or RISC-V. Could also be further adjusted for supporting cpp translation with custom dialects.

I tried to keep most of the code to be as original as possible while here are some of the desisions I make:
1. remove translateToCpp out of emitc namespace (but second thought might add it back now).
2. added a moduleOp to `CppEmitter` to load the context for collecting relevant interfaces.
3. Moved some shared functions to CppTranslationUtils. Although for example `printConstantOp()` is only used for emitc now, but I think it looks clearer this way.

### CODE REVIEW

File: mlir/include/mlir/Target/Cpp/CppEmitter.h [hunk 1]
```diff
@@ -14,21 +14,15 @@
 #define MLIR_TARGET_CPP_CPPEMITTER_H
 
 #include "mlir/IR/BuiltinTypes.h"
-#include "mlir/IR/Value.h"
-#include "llvm/ADT/ScopedHashTable.h"
-#include "llvm/Support/raw_ostream.h"
-#include <stack>
 
 namespace mlir {
-namespace emitc {
```
Comment (joker-eph) [inline mlir/include/mlir/Target/Cpp/CppEmitter.h:23]: Why losing the namespace here?

File: mlir/include/mlir/Target/Cpp/CppTranslationInterface.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_TARGET_CPP_CPPTRANSLATIONINTERFACE_H
+#define MLIR_TARGET_CPP_CPPTRANSLATIONINTERFACE_H
+
+#include "mlir/IR/DialectInterface.h"
+#include "mlir/Support/LogicalResult.h"
+
+namespace mlir {
```
Comment (joker-eph) [inline mlir/include/mlir/Target/Cpp/CppTranslationInterface.h:19]: This is all part of EmitC still right? If so let's keep the namespace.

File: mlir/include/mlir/Target/Cpp/CppTranslationUtils.h [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                                 StringRef callee);
+
+LogicalResult printFunctionArgs(CppEmitter &emitter, Operation *functionOp,
+                                ArrayRef<Type> arguments);
+
+LogicalResult printFunctionArgs(CppEmitter &emitter, Operation *functionOp,
+                                Region::BlockArgListType arguments);
+
+LogicalResult printFunctionBody(CppEmitter &emitter, Operation *functionOp,
+                                Region::BlockListType &blocks);
```
Comment (joker-eph) [inline mlir/include/mlir/Target/Cpp/CppTranslationUtils.h:83]: All of the public APIs should be documented

File: mlir/include/mlir/Target/Cpp/CppTranslationUtils.h [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+#define MLIR_TARGET_CPP_CPPTRANSLATIONUTILS_H
+
+#include "mlir/Dialect/EmitC/IR/EmitC.h"
+#include "mlir/Target/Cpp/TranslateToCpp.h"
+
+using namespace mlir;
+
+/// Convenience functions to produce interleaved output with functions returning
+/// a LogicalResult. This is different than those in STLExtras as functions used
+/// on each element doesn't return a string.
```
Comment (joker-eph) [inline mlir/include/mlir/Target/Cpp/CppTranslationUtils.h:24]: I don't get this comment, the `interleaveComma` in STLExtras takes a functor that does not return a string either

File: mlir/include/mlir/Target/Cpp/CppTranslationUtils.h [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                                         NullaryFunctor betweenFn) {
+  return interleaveWithError(c.begin(), c.end(), eachFn, betweenFn);
+}
+
+template <typename Container, typename UnaryFunctor>
+inline LogicalResult interleaveCommaWithError(const Container &c,
+                                              raw_ostream &os,
+                                              UnaryFunctor eachFn) {
+  return interleaveWithError(c.begin(), c.end(), eachFn, [&]() { os << ", "; });
+}
```
Comment (joker-eph) [inline mlir/include/mlir/Target/Cpp/CppTranslationUtils.h:55]: This does not belong here, it seems like a generic MLIR utility.

File: mlir/include/mlir/Target/Cpp/CppEmitter.h (hunk 1 above)
Comment (weizchen) [reply mlir/include/mlir/Target/Cpp/CppEmitter.h:23]: oh I thought it might be good to let translateToCpp be shared with other dialects (not "owned" by emitc). But now a second thought appears that emitc is the end goal for the cpp translation so I could also add it back to keep it original.

File: mlir/include/mlir/Target/Cpp/TranslateToCpp.h [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+#include "mlir/Target/Cpp/CppEmitter.h"
+#include "mlir/Target/Cpp/CppTranslationInterface.h"
+#include "llvm/ADT/ScopedHashTable.h"
+#include "llvm/Support/raw_ostream.h"
+#include <stack>
+
+namespace mlir {
+
+/// Emitter that uses dialect specific emitters to emit C++ code.
+struct CppEmitter {
```
Comment (joker-eph) [inline mlir/include/mlir/Target/Cpp/TranslateToCpp.h:28]: Please make this a class, let's keep "struct" for pure trivial data storage.

File: mlir/include/mlir/Target/Cpp/CppTranslationUtils.h (hunk 4 above)
Comment (weizchen) [reply mlir/include/mlir/Target/Cpp/CppTranslationUtils.h:24]: This part actually comes from the original EmitC code, I tried to keep most of the code as original as possible but I could double check.

File: mlir/include/mlir/Target/Cpp/CppTranslationUtils.h [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_TARGET_CPP_CPPTRANSLATIONUTILS_H
+#define MLIR_TARGET_CPP_CPPTRANSLATIONUTILS_H
+
+#include "mlir/Dialect/EmitC/IR/EmitC.h"
+#include "mlir/Target/Cpp/TranslateToCpp.h"
+
+using namespace mlir;
```
Comment (joker-eph) [inline mlir/include/mlir/Target/Cpp/CppTranslationUtils.h:20]: That should be in any header.

**Comment by mgehre-amd** [comment]
As I understand the current design of emitc, it's intended that the translator will only accept the emitc dialect, and other dialects are instead supposed to convert into emitc before translation.

Your PR wants to move away from that design, correct?

**Comment by weizchen** [comment]
> As I understand the current design of emitc, it's intended that the translator will only accept the emitc dialect, and other dialects are instead supposed to convert into emitc before translation.
> 
> Your PR wants to move away from that design, correct?

Hi, thank you for your comment. I also thought about this before but I think not necessarily, I would not say this PR will conflict with the current design of emitc. 

Although in this PR, dialects like func getting separated out seems like we want to do direct translation to cpp, but as emitc func gets more functionality, we could still remove func and only use emitc. I think this will be more useful as people aim for a more customized cpp translation in the future.

**Comment by marbre** [comment]
Thank you for your work @weizchen. It is undeniable that the emitter can be improved. However, after a first glance, I think there are both conceptual and technical issues. I'm afraid that this will definitely require some discussions and iterations before it could potentially be merged.

Let me provide more context and some background information. When we proposed the emitter for upstream inclusion, the were concerns and push back mainly due to the API of the emitter. At that time, the consensus was to keep the surface of the emitter as small as possible. This also made it necessary for us to fork the emitter for our use case in IREE. However, we have now evolved the EmitC dialect to a point where we no longer need a forked emitter, as we can achieve what we need through conversions. Hence, @simon-camp recently removed the forked emitter from IREE with PR [#16357@IREE](https://github.com/openxla/iree/pull/16357).

The intention is to remove all non-EmitC operations from the emitter. Support for operations of the SCF dialect were removed with PR #68206 and we recently accepted PR #83798 that removed support for the Arith dialect. With PR #79612, I added a func, call and return operation. The only reason for us not yet removing support for the Func dialects ops from emitter is to not break downstream users as EmitC's operations don't support multiple return values. Instead of extending the emitter, we aim to further develop the dialect in a way that everything can be accomplished through conversions. To me, it seems that the proposed changes do not align with this intention.
For completeness, this also seems related to the discussion [[RFC] The Natural Intersection of MLIR and C](https://discourse.llvm.org/t/rfc-the-natural-intersection-of-mlir-and-c/68708/2).

Furthermore, there are some mechanical issues. For example the implementation in `CppTranslationUtils` heavily relies on EmitC ops and further pulls in the CF dialect. Hence, it not just provides shared functions as most functions are not *dialect independent*.

**Comment by weizchen** [comment]
Hi @marbre, thank you so much for this detailed explanation and references! I totally agree that discussion is needed on this PR's implementation. 

>  Instead of extending the emitter, we aim to further develop the dialect in a way that everything can be accomplished through conversions. To me, it seems that the proposed changes do not align with this intention.

I also discover the trend of adding more conversions to EmitC to remove direct dialect translations which totally makes sense to me. I agree that separating out dialects like Func in this PR does not align with the current emitc strategy but what I vision for this is we have a fully supported emitter from EmitC Dialect, all main dialects are first translated to EmitC and then to C code. While the `Dialects` folder is not intended for Func or CF dialects shown in this PR, but for dialects like ArmSVE or X86Vector that will require more specialized translation.

> Furthermore, there are some mechanical issues. For example the implementation in CppTranslationUtils heavily relies on EmitC ops and further pulls in the CF dialect. Hence, it not just provides shared functions as most functions are not dialect independent.

And for this, I really struggled creating such Utils file, especially as the translations are all moving towards the EmitC Dialect. The original idea was to use this file for similar op translations to avoid code duplication (like arith constant and emitc constant), and I want to keep `EmitCToCppTranslation.cpp` purely containing `printOperation`. But you are right, making these functions living with EmitC translation would make sense.

This PR actually originated from a need to translate our customized dialect to c++ where we found current emitc structure is a bit hard to inherit and reuse. So we think that with this, cpp translation might be more sustainable and extensible. I wonder what do you think of this, your idea is highly valued. Thanks!

**Comment by joker-eph** [comment]
> This PR actually originated from a need to translate our customized dialect to c++ where we found current emitc structure is a bit hard to inherit and reuse. So we think that with this, cpp translation might be more sustainable and extensible. I wonder what do you think of this, your idea is highly valued. Thanks!

It would be interesting to understand why you can't use the dialect conversion mechanism to lower to EmitC actually: can you come up with examples?
The solution may also be to extend EmitC concepts to better support the missing aspects that you found.

**Comment by weizchen** [comment]
> It would be interesting to understand why you can't use the dialect conversion mechanism to lower to EmitC actually: can you come up with examples? The solution may also be to extend EmitC concepts to better support the missing aspects that you found.

Hi, thank you for the response and your previous reviews! Regarding the question, one reason is the dialect is designed to fully target its C API, so going through EmitC adds one layer of overhead as things get a bit complicated to represent customized C operations, types and attributes with opaque families. I like the opaque and verbatim ops as they solve issues but I cannot really do much with them. For example like a compound type or like `<Cache_Level::L2, 2>` with Cache Level is a enum attribute and 2 is a variable, this is something I found a bit headache. But as you said, I think it is possible to convert to EmitC but the solution might not be as straightforward as extending the emitter.

**Comment by joker-eph** [comment]
> But as you said, I think it is possible to convert to EmitC but the solution might not be as straightforward as extending the emitter.

I suspect we'd want to understand this with much more details, with in-tree examples, before opening the door to bespoke emitters. As I mentioned before "the solution may also be to extend EmitC concepts", but I think we won't know about this if we just expose interfaces without first building a good understanding of the limits.

**Comment by weizchen** [comment]
> I think we won't know about this if we just expose interfaces without first building a good understanding of the limits

Yeah sorry for not making this clear. So for example, the compound type I mentioned before, `<Cache_Level::L2, 2>`, it could be an exhaustive process to represent by opaque types, but if we can extend the printType function; or we add a printCustomType function and call both in the emitter, only a few lines of code can cover all the situations. This example is actually a bad example that can reveal the limitation. This is because if we are going to extend the member function in the CppEmitter class, things will get a bit messy to deal with the interface function, so a type interface is preferred in this situation to solve the problem. But with this interface, it provides a way for external dialects to register cpp translation into the emitc.

**Comment by marbre** [comment]
I think it would be really valuable if you could post an IR snippet @weizchen. It is of course possible, and even likely, that we're missing some ops in the EmitC dialect. It's developed and extended on a need-driven basis. However, as @joker-eph mentioned, it would be helpful to better understand this before exposing the interfaces. Maybe there is a convenient way using conversions that opens up if we add some more missing peaces to the dialect.

**Comment by weizchen** [comment]
> I think it would be really valuable if you could post an IR snippet

Sure thing! Here is an IR snippet that has a similar idea with the Cache_Level example I talked about.

This TQue type comprises one enum attribute: TPosition (VECIN, VECOUT...) that represents hardware positions and a buffer number. 

``` mlir
// void type1(TQue<QuePosition::VECIN, 1> que1) {
func.func @type1(%arg0: !ascendc.TQue<VECIN, 1>) {
    return
}

// void type2(TQue<QuePosition::VECIN, 2> que1) {
func.func @type2(%arg0: !ascendc.TQue<VECIN, 2>) {
    return
}

// void type3(TQue<QuePosition::VECIN, 3> que1) {
func.func @type3(%arg0: !ascendc.TQue<VECIN, 3>) {
    return
}

// void type4(TQue<QuePosition::VECOUT, 2> que1) {
func.func @type4(%arg0: !ascendc.TQue<VECOUT, 2>) {
    return
}

// void type5(TQue<QuePosition::A1, 2> que1) {
func.func @type5(%arg0: !ascendc.TQue<A1, 2>) {
    return
}

// void type6(TQue<QuePosition::C1, 2> que1) {
func.func @type6(%arg0: !ascendc.TQue<C1, 2>) {
    return
}

```

Please correct me if I am missing anything, but in my understanding, if we want to lower this TQue type to EmitC Dialect, we have to lower it to opaque type one by one. So for the six types above, I would need six different conversions to opaque, while this number can greatly increase with different combinations of TPosition and buffer number.

But if we can support custom printing, for example like `os << "TQue<QuePosition::" << tque.getTposition() << ", " << tque.getBufNum() << ">";` this would easily solve the problem. I hope this brings more context to you and I am very open to your suggestions about EmitC Conversion.

**Comment by marbre** [comment]
> Suppose TQue has some member functions that we want to call. For example, in mlir, we have `%tensor1 = ascendc.alloc_tensor(%QueueX) : !ascendc.TQue<VECIN, 2> -> !ascendc.LocalTensor<16xf16>` which will be translated to `LocalTensor<half> tensor1 = que1.AllocTensor<half>();` Do we have some good representation for this `.` or `->` for the member function? Thanks in advance!

Unfortunately, there is no convenient way to express a member access operator. While this could be added to the dialect, we hesitated to do so, as this is closely related to an lvalue/rvalue discussion we're currently facing (@simon-camp will post an update to this later on discourse). Alongside we would need to add function pointers (related to `func.constant`) and an operation resembling `func.call_indirect`. While adding these two to the dialect is straight forward, it requires some additional work in the emitter.

TL;DR, right know there is no convenient way, but you could use `emtic.verbatim` to emit a macro (if you want to define it via MLIR and don't want to provide a hand-written header file) and together with using `emitc.call_opaque`.

**Comment by simon-camp** [comment]
> Unfortunately, there is no convenient way to express a member access operator. While this could be added to the dialect, we hesitated to do so, as this is closely related to an lvalue/rvalue discussion we're currently facing (@simon-camp will post an update to this later on discourse). Alongside we would need to add function pointers (related to `func.constant`) and an operation resembling `func.call_indirect`. While adding these two to the dialect is straight forward, it requires some additional work in the emitter.

Here is the mentioned [discourse thread](https://discourse.llvm.org/t/rfc-separate-variables-from-ssa-values-in-emitc/75224).

**Comment by weizchen** [comment]
I see, thank you all for the help. I will adopt the EmitC conversion to see how it goes. For the purpose of this PR, I think the extension could always be rediscussed if more need comes up for customization in the future if necessary. For now I will close this PR, thanks again for the info and the help!

**Comment by marbre** [comment]
> I see, thank you all for the help. I will adopt the EmitC conversion to see how it goes. For the purpose of this PR, I think the extension could always be rediscussed if more need comes up for customization in the future if necessary. For now I will close this PR, thanks again for the info and the help!

Thanks for the discussion! Feel free to open up topics at [discourse](https://discourse.llvm.org/c/mlir/31) for any further EmitC related discussions or requests.
